### DISCUSSION TITLE
How is Line Color picked?

### DISCUSSION THREAD

**Discussion by dibli-goost** [bnpr/Malt#458] [Q&A] [answered]
With my setup, Line Color is derived by darkening the shadow pass of my shader, and passing it into the Line Color output.

It looks like so:
![image](https://user-images.githubusercontent.com/31767887/207153967-d867d58a-d075-43bf-8d55-c5f182c250dc.png)

However, I've sorta run into an issue with this, which is that there is some ambiguity to Malt's choice of Line color.

![image](https://user-images.githubusercontent.com/31767887/207153352-cdd8944c-5bdf-458b-9d7a-63bcad419891.png)

When a line sits on a space which is inhabited by multiple colors, it has to pick between them. And sometimes it doesn't seem to choose the right one. On this striped shirt here, it can't really decide whether it wants the lines to be grey or blackish-red. Of course, this could be prevented by me creating a dedicated line color setup, where I just manually set it, but I'm lazy 😛.

How exactly does Malt decide that color? Is it based off of depth or something? As I twist the camera around, I can see that the color will shift depending upon angle. Maybe it picks the color which is further down the screen?

![image](https://user-images.githubusercontent.com/31767887/207155475-367cd0b6-473b-40fc-ae9b-954d3f8349ee.png)

---------------

Wondering if there could be a way to push Malt to pick a specific one over the other, to get more consistent results. To say maybe, "pick the color with the lowest value", or maybe "pick the color that is more saturated". It might not be as great as I'm thinking it would be though. The current way that it picks works well on areas with less drastic color shifts, like with the skin here. 

![image](https://user-images.githubusercontent.com/31767887/207156370-44ffa2b5-73c2-4744-92d1-fe87aa9deb1c.png)

It's also much harder to catch if the lines are smaller too. With the size i usually use, it's kinda hard to notice, but it has been kinda bothering me for a bit.

**Comment by pragma37** [comment, marked as answer]
> How exactly does Malt decide that color? Is it based off of depth or something?

Essentially, yes. The line nearest to the camera "wins".

Lines are detected by checking for different types of boundaries with nearby pixels.
So that means if pixel A detects a line when checking against pixel B, then pixel B will also detect a line when checking against pixel A.
So unless we discard one of the 2 cases, we end up with 2 contiguous parallel lines everywhere, which more often than not causes unintended results.

I think your best solution here would be to make sure both materials output the same line color.
Using vertex colors or a texture should do the trick.

It should be possible to implement some sort of line priority system, but I don't think this should be a built-in feature.
